explorer/interpreter/interpreter.cpp

```diff
@@ -860,6 +865,42 @@ auto Interpreter::Convert(Nonnull<const Value*> value,
       }
       return Convert(value, destination_type, source_loc);
     }
+    case Value::Kind::PointerValue: {
+      if (destination_type->kind() != Value::Kind::PointerType ||
+          cast<PointerType>(destination_type)->type().kind() !=
+              Value::Kind::NominalClassType) {
+        // No conversion needed.
+        return value;
+      }
+
+      // Get pointed value.
+      const auto* src_ptr = cast<PointerValue>(value);
+      CARBON_ASSIGN_OR_RETURN(const auto* pointee,
+                              heap_.Read(src_ptr->address(), source_loc))
+      if (pointee->kind() == Value::Kind::NominalClassType) {
+        // TODO: When does that happen? Just return as-is.
+        return value;
+      }
+
+      const auto* dest_ptr = cast<PointerType>(destination_type);
+      std::optional<Nonnull<const NominalClassValue*>> class_subobj =
+          cast<NominalClassValue>(pointee);
+      auto new_addr = src_ptr->address();
+      while (class_subobj) {
+        if (TypeEqual(&class_subobj.value()->type(), &dest_ptr->type(),
```

Comment:
Would it work to use `destination_type` instead of `dest_ptr->type()` here? I think we should be cautious about accessing the type of a run-time `Value` in the interpreter, because the compiler won't be able to do that.

Reply:
Not really unfortunately, `destination_type` would just tell us it is a pointer in general, but we need to match the pointee type with the destination pointer type.

Reply:
Ah, sorry, I misunderstood what this code was doing. In my defense, `PointerType::type` is very confusingly-named; I've sent #2468 to fix that.